components/tikv_util/src/metrics/mod.rs

```diff
@@ -36,16 +46,63 @@ mod metrics_reader;
 
 pub type RecordPairVec = Vec<pdpb::RecordPair>;
 
-pub fn dump(should_simplify: bool) -> String {
+lazy_static! {
+    // the registry for high priority metrics.
+    pub static ref HIGH_PRIORITY_REGISTRY: Registry = Registry::new();
+}
+
+static METRICS_COMPACT_LEVZEL: AtomicU32 = AtomicU32::new(MetricsCompactLevel::NoCompact as u32);
+static METRICS_LEVEL: AtomicU32 = AtomicU32::new(MetricsLevel::All as u32);
+
+// reduce the return frequency of normal metrics factor
+const NORMAL_METRICS_REDUCE_FACTOR: u64 = 2;
+static METRICS_REQUEST_COUNTER: AtomicU64 = AtomicU64::new(0);
+
+pub fn set_metrics_compact_level(level: MetricsCompactLevel) {
+    METRICS_COMPACT_LEVZEL.store(level as u32, Ordering::Relaxed);
+}
+
+pub fn get_metrics_compact_level() -> MetricsCompactLevel {
+    METRICS_COMPACT_LEVZEL.load(Ordering::Relaxed).into()
+}
+
+pub fn set_metrics_level(level: MetricsLevel) {
+    METRICS_LEVEL.store(level as u32, Ordering::Relaxed);
+}
+
+pub fn get_metrics_level() -> MetricsLevel {
+    METRICS_LEVEL.load(Ordering::Relaxed).into()
+}
+
+pub fn should_return_normal_metrics() -> bool {
+    let metrics_level = get_metrics_level();
+    metrics_level == MetricsLevel::All
+        || (metrics_level == MetricsLevel::ReduceFrequency
+            && (METRICS_REQUEST_COUNTER.load(Ordering::Relaxed) % NORMAL_METRICS_REDUCE_FACTOR
+                == 0))
+}
+
+pub fn dump() -> String {
     let mut buffer = vec![];
-    dump_to(&mut buffer, should_simplify);
+
+    dump_to(&mut buffer);
     String::from_utf8(buffer).unwrap()
 }
 
-pub fn dump_to(w: &mut impl Write, should_simplify: bool) {
+pub fn dump_to(w: &mut impl Write) {
+    METRICS_REQUEST_COUNTER.fetch_add(1, Ordering::Release);
+
+    dump_metrics_to(w, HIGH_PRIORITY_REGISTRY.gather());
+    if should_return_normal_metrics() {
+        dump_metrics_to(w, prometheus::gather());
+    }
+}
+
+pub fn dump_metrics_to(w: &mut impl Write, metric_families: Vec<MetricFamily>) {
+    let simplify_level = get_metrics_compact_level();
+
     let encoder = TextEncoder::new();
-    let metric_families = prometheus::gather();
-    if !should_simplify {
+    if simplify_level == MetricsCompactLevel::NoCompact {
```

Comment:
Why need NoCompact? seems this also may ignore high-priority-metrics when the simplify_level is other values?

Reply:
`NoCompact` means just returning the original data, in this branch, it calls `encoder.encode` for all metrics at L106.